pkg/controller/atlasconnection/atlasconnection_controller.go

```diff
@@ -0,0 +1,409 @@
+	ptr "k8s.io/utils/pointer"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	dbaasv1alpha1 "github.com/RHEcosystemAppEng/dbaas-operator/api/v1alpha1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/dbaas"
+	mdbv1 "github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1/status"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/atlas"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/customresource"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/watch"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/workflow"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/util/kube"
+	"go.mongodb.org/atlas/mongodbatlas"
+)
+
+const (
+	DBUserNameKey = "username"
+	DBPasswordKey = "password"
+
+	digits   = "0123456789"
+	specials = "~=+%^*/()[]{}/!@#$?|"
+	all      = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz" + digits + specials
+)
+
+// MongoDBAtlasConnectionReconciler reconciles a MongoDBAtlasConnection object
+type MongoDBAtlasConnectionReconciler struct {
+	Client    client.Client
+	Clientset *kubernetes.Clientset
+	watch.ResourceWatcher
+	Log             *zap.SugaredLogger
+	Scheme          *runtime.Scheme
+	AtlasDomain     string
+	GlobalAPISecret client.ObjectKey
+	EventRecorder   record.EventRecorder
+}
+
+// Dev note: duplicate the permissions in both sections below to generate both Role and ClusterRoles
+
+// +kubebuilder:rbac:groups=atlas.mongodb.com,resources=MongoDBAtlasConnections,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=atlas.mongodb.com,resources=MongoDBAtlasConnections/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch
+
+// +kubebuilder:rbac:groups=atlas.mongodb.com,namespace=default,resources=MongoDBAtlasConnections,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=atlas.mongodb.com,namespace=default,resources=MongoDBAtlasConnections/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",namespace=default,resources=secrets,verbs=get;list;watch
+
+func (r *MongoDBAtlasConnectionReconciler) Reconcile(cx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = cx
+	log := r.Log.With("MongoDBAtlasConnection", req.NamespacedName)
+
+	conn := &dbaas.MongoDBAtlasConnection{}
+	if err := r.Client.Get(cx, req.NamespacedName, conn); err != nil {
+		if errors.IsNotFound(err) {
+			// CR deleted since request queued, child objects getting GC'd, no requeue
+			log.Info("MongoDBAtlasConnection resource not found, has been deleted")
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "Error fetching MongoDBAtlasConnection for reconcile")
+		return ctrl.Result{}, err
+	}
+
+	if isReadyForBinding(conn) {
```

Comment:
not sure this is the right question.  what if the same CR updated with a different instance id? in that case it will return an old instance connection ?

Reply:
Good point. I removed this check to avoid this side effect :)

Reply:
So, as discussed, we need that check, else it will goes and create a new secret every time and update the status.
we can do something like if `isReadyForBinding=true` and `instanceid=status.instance-id`, then return nil, else delete the existing secret/ create a new secret and update status
but currently, we are not saving the instance id in status. WDYT?

Reply:
Yes we need to address this issue. We can save the instanceID and any other needed info in Status.ConnectionInfo map, so that we can use it for later checking.